src/lib.rs

```diff
@@ -1923,16 +1923,32 @@ impl Connection {
 
         hdr.to_bytes(&mut b)?;
 
-        // Make sure we have enough space left for the header, the payload
+        // Calculate the size of header including the payload
         // length, the packet number and the AEAD overhead.
-        left = left
-            .checked_sub(b.off() + pn_len + overhead)
-            .ok_or(Error::Done)?;
+        let mut hdr_len = b.off() + pn_len + overhead;
 
         // We assume that the payload length, which is only present in long
         // header packets, can always be encoded with a 2-byte varint.
         if pkt_type != packet::Type::Short {
-            left = left.checked_sub(2).ok_or(Error::Done)?;
+            hdr_len += 2;
+        }
+
+        // Make sure we have enough space left for packet payload.
+        if let Some(payload_left) = left.checked_sub(hdr_len) {
+            left = payload_left;
+        } else {
+            // When there is nothing to write, we stil need to update
+            // app_limited flag before return.
+            //
+            // This usually happens when we called send() multiple times
+            // in a single event to send a large data.
+            // When cwnd is almost full and the last send() can't send
+            // a new packet because the packet header length is bigger
+            // than the allowed length by cwnd, it's considered as a
+            // full of cwnd is sent and app_limited is set to
+            // false to make cwnd grow when ACK is received.
```

Comment:
I'm finding it a bit hard to comprehend this comment

Reply:
I updated the comment. Let me know if it works.

Reply:
much clearer, thanks!